### PR TITLE
New feature : Variables

### PR DESCRIPTION
**Pull request by NicolasEYSSERIC**

### CODE REVIEW

File: core-api/src/main/java/org/silverpeas/core/variables/VariablePeriod.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+  public Variable getVariable() {
+    return variable;
+  }
+
+  public String getValueForHTML() {
+    return WebEncodeHelper.javaStringToHtmlParagraphe(getValue());
+  }
+
+  public VariablePeriod setId(String id) {
```
Comment (silveruser) [inline core-api/src/main/java/org/silverpeas/core/variables/VariablePeriod.java]: MAJOR Add the "@Override" annotation above this method signature [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1161)

File: core-api/src/main/java/org/silverpeas/core/variables/VariablePeriod.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+import java.util.Date;
+
+@Entity
+@Table(name = "sb_variables_period")
+@NamedQueries({
+    @NamedQuery(name = "currentPeriods", query = "select p from VariablePeriod p where p" +
+        ".startDate <= :today and :today <= p.endDate order by p.variable.label ASC")})
+public class VariablePeriod extends SilverpeasJpaEntity<VariablePeriod, UuidIdentifier>
```
Comment (silveruser) [inline core-api/src/main/java/org/silverpeas/core/variables/VariablePeriod.java]: MINOR Override the "equals" method in this class. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS2160)

File: core-web/src/main/java/org/silverpeas/core/webapi/variables/VariableEntity.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    entity.setValue(variable.getCurrentPeriod().getValue());
+    return entity;
+  }
+
+  public Variable toVariable() {
+    return new Variable(label, description);
+  }
+
+  public List<VariablePeriod> toPeriods() {
+    List<VariablePeriod> periods = new ArrayList<>();
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/webapi/variables/VariableEntity.java]: MAJOR Rename "periods" which hides the field declared at line 49. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: core-api/src/main/java/org/silverpeas/core/variables/Variable.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import javax.persistence.Table;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+@Entity
+@Table(name = "sb_variables_variable")
+@NamedQueries({
+    @NamedQuery(name = "allVariables", query = "select v from Variable v order by v.label ASC")})
+public class Variable extends SilverpeasJpaEntity<Variable, UuidIdentifier> implements
```
Comment (silveruser) [inline core-api/src/main/java/org/silverpeas/core/variables/Variable.java]: MINOR Override the "equals" method in this class. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS2160)

File: core-api/src/main/java/org/silverpeas/core/variables/Variable.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  public void setLabel(final String label) {
+    this.label = label;
+  }
+
+  public void setDescription(final String description) {
+    this.description = description;
+  }
+
+  public Variable setId(String id) {
```
Comment (silveruser) [inline core-api/src/main/java/org/silverpeas/core/variables/Variable.java]: MAJOR Add the "@Override" annotation above this method signature [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1161)

File: core-library/src/main/java/org/silverpeas/core/variables/DefaultVariablesRepository.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package org.silverpeas.core.variables;
+
+import org.silverpeas.core.persistence.datasource.repository.jpa.NamedParameters;
+import org.silverpeas.core.persistence.datasource.repository.jpa.SilverpeasJpaEntityRepository;
+
+import javax.inject.Singleton;
+import java.util.Date;
```
Comment (silveruser) [inline core-library/src/main/java/org/silverpeas/core/variables/DefaultVariablesRepository.java]: MINOR Remove this unused import 'java.util.Date'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core-war/src/main/java/org/silverpeas/web/variables/VariablesRequestRouter.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+import org.silverpeas.core.admin.user.model.User;
+import org.silverpeas.core.util.SilverpeasList;
+import org.silverpeas.core.util.StringUtil;
+import org.silverpeas.core.web.http.HttpRequest;
+import org.silverpeas.core.web.mvc.controller.ComponentContext;
+import org.silverpeas.core.web.mvc.controller.MainSessionController;
+import org.silverpeas.core.web.mvc.route.ComponentRequestRouter;
+
+public class VariablesRequestRouter
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/variables/VariablesRequestRouter.java]: MAJOR This class has 6 parents which is greater than 5 authorized. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)

**Comment by silveruser** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core-war/src/main/java/org/silverpeas/web/variables/VariablesRequestRouter.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+import org.silverpeas.core.admin.user.model.User;
+import org.silverpeas.core.util.SilverpeasList;
+import org.silverpeas.core.util.StringUtil;
+import org.silverpeas.core.web.http.HttpRequest;
+import org.silverpeas.core.web.mvc.controller.ComponentContext;
+import org.silverpeas.core.web.mvc.controller.MainSessionController;
+import org.silverpeas.core.web.mvc.route.ComponentRequestRouter;
+
+public class VariablesRequestRouter extends ComponentRequestRouter<VariablesSessionController> {
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/variables/VariablesRequestRouter.java]: MAJOR This class has 6 parents which is greater than 5 authorized. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)

**Comment by silveruser** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by NicolasEYSSERIC** [comment]
Do not forget to check out https://github.com/Silverpeas/Silverpeas-Components/pull/566

**Comment by mmoqui** [comment]
This PR is closed and replaced by the PR https://github.com/Silverpeas/Silverpeas-Core/pull/868

**Pull request closed without merging**